benches/dapol.rs

```diff
@@ -1,175 +1,250 @@
-        let items = build_item_list(num_leaves, tree_height);
-        let mut rng = thread_rng();
-        let item_range = Uniform::new(0usize, num_leaves);
-
-        let dapol = build_dapol_tree::<blake3::Hasher, RangeProofSplitting>(&items, tree_height);
-        group.bench_function(BenchmarkId::new("splitting", tree_height), |bench| {
-            bench.iter_batched(
-                || {
-                    // generate a proof
-                    let item_idx = rng.sample(item_range);
-                    let tree_index = &items[item_idx].0;
-                    (item_idx, dapol.generate_proof(tree_index).unwrap())
-                },
-                |(item_idx, proof)| {
-                    // time proof verification
-                    proof.verify(&dapol.root(), &items[item_idx].1.get_proof_node())
-                },
-                BatchSize::SmallInput,
-            );
-        });
-
-        let dapol = build_dapol_tree::<blake3::Hasher, RangeProofPadding>(&items, tree_height);
-        group.bench_function(BenchmarkId::new("padding", tree_height), |bench| {
-            bench.iter_batched(
-                || {
-                    // generate a proof
-                    let item_idx = rng.sample(item_range);
-                    let tree_index = &items[item_idx].0;
-                    (item_idx, dapol.generate_proof(tree_index).unwrap())
-                },
-                |(item_idx, proof)| {
-                    // time proof verification
-                    proof.verify(&dapol.root(), &items[item_idx].1.get_proof_node())
-                },
-                BatchSize::SmallInput,
-            );
-        });
+    thread_counts.push(max_thread_count);
+
+    let mut ndm_smt = Option::<NdmSmt>::None;
+
+    e.advance().unwrap();
+
+    for h in TREE_HEIGHTS.into_iter() {
+        e.advance().unwrap();
+
+        for t in thread_counts.iter() {
+            e.advance().unwrap();
+
+            for u in NUM_USERS.into_iter() {
+                e.advance().unwrap();
+
+                let max_users_for_height = 2_u64.pow((h - 1) as u32);
+
+                if u > max_users_for_height {
+                    break;
+                }
+
+                let tup: (Height, MaxThreadCount, u64) =
+                    (Height::from(h), MaxThreadCount::from(*t), u);
+
+                // tree build compute time
+                group.bench_function(
+                    BenchmarkId::new(
+                        "build_tree",
+                        format!("{:?}/{:?}/NUM_USERS: {:?}", &tup.0, &tup.1, &tup.2),
+                    ),
+                    |bench| {
+                        bench.iter(|| {
+                            e.advance().unwrap();
+                            ndm_smt = Some(setup::build_ndm_smt(tup.clone()));
+                        });
+                    },
+                );
+
+                // tree build memory usage
+                let alloc = alloc.read().unwrap();
```

Comment:
You have to take 2 readings to get the memory usage of the ndm_smt variable: 1 before construction and 1 after, then take the diff. I'd say this should give an accurate reading:
```rust
for ...
  for ...
    for ...
      e.advance().unwrap();
      let before = alloc.read().unwrap();

      group.bench_function(...);

      e.advance().unwrap();
      let after = alloc.read().unwrap();

      // mem used is the difference between the 2 measurements
      let diff = after - before;
```

Reply:
I am getting very strange memory usage readings after making these changes, (check out that last line):

```bash
     Running benches/dapol.rs (target/release/deps/dapol-c607f29cfb9c0202)
Benchmarking dapol/build_tree/Height(16)/MaxThreadCount(4)/NUM_USERS: 10000: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 7.3s.
dapol/build_tree/Height(16)/MaxThreadCount(4)/NUM_USERS: 10000
                        time:   [784.02 ms 807.24 ms 824.34 ms]
                        change: [-1.3000% +2.6972% +6.9786%] (p = 0.24 > 0.05)
                        No change in performance detected.
Found 2 outliers among 10 measurements (20.00%)
  2 (20.00%) low severe

Metrics { variable: TreeBuild, mem_usage: "13.69 MB", file_size: "2.06 MB" }

Benchmarking dapol/build_tree/Height(16)/MaxThreadCount(4)/NUM_USERS: 20000: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 8.9s.
dapol/build_tree/Height(16)/MaxThreadCount(4)/NUM_USERS: 20000
                        time:   [1.0546 s 1.0895 s 1.1233 s]
                        change: [-0.9219% +4.1641% +9.6951%] (p = 0.15 > 0.05)
                        No change in performance detected.

Metrics { variable: TreeBuild, mem_usage: "3.19 MB", file_size: "4.08 MB" }

Benchmarking dapol/build_tree/Height(16)/MaxThreadCount(4)/NUM_USERS: 30000: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 9.3s.
dapol/build_tree/Height(16)/MaxThreadCount(4)/NUM_USERS: 30000
                        time:   [1.0532 s 1.0714 s 1.0889 s]
                        change: [-0.1598% +5.6191% +13.358%] (p = 0.11 > 0.05)
                        No change in performance detected.

Metrics { variable: TreeBuild, mem_usage: "10.33 MB", file_size: "6.11 MB" }

Benchmarking dapol/build_tree/Height(16)/MaxThreadCount(8)/NUM_USERS: 10000: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 10.5s.
dapol/build_tree/Height(16)/MaxThreadCount(8)/NUM_USERS: 10000
                        time:   [1.4478 s 1.4533 s 1.4588 s]
                        change: [-0.7875% -0.3223% +0.2068%] (p = 0.26 > 0.05)
                        No change in performance detected.
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) low mild
  1 (10.00%) high mild

Metrics { variable: TreeBuild, mem_usage: "16777216.00 TB", file_size: "2.06 MB" }
```